generate_workspace/src/test/java/com/google/devtools/build/workspace/maven/VersionResolverTest.java

```diff
@@ -0,0 +1,108 @@
+import org.junit.runners.JUnit4;
+import org.mockito.Mockito;
+
+/**
+ * Tests for {@link VersionResolver}.
+ */
+//TODO(petros): Test this actually selects highest version. VersionRangeResult is final class so Mockito cannot mock it.
+@RunWith(JUnit4.class)
+public class VersionResolverTest {
+
+  /**
+   * Ensures that an exception is thrown if there is a version range resolution exception.
+   * This occurs when aether is unable to resolve the version.
+   */
+  @Test
+  public void failsOnResolutionException() {
+    RepositorySystem system = Mockito.mock(RepositorySystem.class);
+    try {
+      Mockito.when(system.resolveVersionRange(anySession(), anyRangeRequest()))
+          .thenThrow(new VersionRangeResolutionException(any()));
+      Aether aether = Aether.builder().systemSession(anySession(), system).build();
+      VersionResolver resolver = new VersionResolver(aether);
+
+      resolver.resolveVersion("something", "something", "1.0");
+      fail();
+    } catch (VersionRangeResolutionException e) {
+      // This should be caught by the VersionResolver
+      fail();
+    } catch (ArtifactBuilder.InvalidArtifactCoordinateException e) {
+      String expected =
+          "Unable to find a version for something:something:1.0 due to Failed to resolve version range";
+      assertThat(e.getMessage()).isEqualTo(expected);
+    }
+  }
+
+  /**
+   * Ensures that an exception is thrown if there is an invalid version range. An invalid version range
+   * is one which is either (1) equal to null or (2) returns null when asked for highest version.
+   */
+  @Test
+  public void failsOnInvalidVersionRange() {
+    RepositorySystem system = Mockito.mock(RepositorySystem.class);
+
+    try {
+      // Using `anyRangeResult()` will ensure that rangeResult.highestVersion() == null.
+      Mockito.when(system.resolveVersionRange(anySession(), anyRangeRequest())).thenReturn(anyRangeResult());
+      Aether aether = Aether.builder().systemSession(anySession(), system).build();
+      VersionResolver resolver = new VersionResolver(aether);
+
+      resolver.resolveVersion("something", "something", "1.0");
+      fail();
+    } catch (VersionRangeResolutionException e) {
+      // This should be caught by the VersionResolver
+      fail();
+    } catch (ArtifactBuilder.InvalidArtifactCoordinateException e) {
+      String expected =
+          "Unable to find a version for something:something:1.0 due to Invalid Range Result";
+      assertThat(e.getMessage()).containsMatch(expected);
+    }
+  }
+
+  /**
+   * Asserts that given a soft pinned version specification, it selects that version,
+   * and does not get the highest version. "3.4" is an example of a soft pinned version specification.
+   */
+  //TODO(petros): implicitly using Maven central for resolving version. This is sketchy.
+  @Test
+  public void softPinnedVersions() {
+    try {
+      String version =
+          defaultResolver().resolveVersion("something", "something", "3.4");
```

Comment:
> Keep in mind that we don't want to test the actual version resolution. That's Aether's business. We just want to be able to test how your code behaves.


I am not we need to add any further tests or whether we can need the soft pin tests. For full code coverage, I believe we simply need to cover the following cases:

* given a null result, it throws an exception
* given a versionrangeresolution exception, it throws an exception
* given neither, it doesn't throw an exception and returns something.

> Have it return a Version. Or a list of versions. You don't need a response, here, just the version.

Again this is a tradeoff. Finer grained tests will mean more responsibility to mimic aether/maven's behavior. 

I assume we are trying to match with how aether or maven resolves versions. If so, returning a list of versions means we will need to include code to _decide_ which version to pick. Right now, we can simply write `versionRangeResult.getHighestVersion()` and call it a day. However, if we return the list, then we will need to account for any weird behavior aether/maven may have. 

For example, Aether internally uses `VersionRangeResult.getHighestVersion()` for resolving all versions. For a soft pin, I have found that this returns the soft pinned value provided it exists. This is regardless of whether the soft pinned value is the highest version or not. The behavior when that soft pin does not exist is undocumented. I am guessing it will be the next closest version, but I am not sure on that front.